src/persist-client/src/read.rs

```diff
@@ -0,0 +1,137 @@
+use std::num::NonZeroUsize;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use differential_dataflow::difference::Semigroup;
+use futures_util::Stream;
+use mz_persist_types::{Codec, Codec64};
+use serde::{Deserialize, Serialize};
+use timely::progress::{Antichain, Timestamp};
+use tracing::warn;
+
+use crate::error::Error;
+
+// WIP This probably doesn't need to be Clone, so I've omitted it for now.
+// Pretty sure we could make that work if necessary.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SnapshotShard(PhantomData<()>);
+
+pub struct SnapshotIter<K, V, T, D>(PhantomData<(K, V, T, D)>);
+
+impl<K, V, T, D> Stream for SnapshotIter<K, V, T, D> {
+    type Item = (K, V, T, D);
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        todo!("{:?}", cx)
+    }
+}
+
+pub struct Tail<K, V, T, D>(PhantomData<(K, V, T, D)>);
+
+impl<K, V, T, D> Stream for Tail<K, V, T, D> {
+    // WIP: We also need a way to delivery progress information from this.
+    type Item = (K, V, T, D);
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        todo!("{:?}", cx)
+    }
+}
+
+pub struct ReadHandle<K, V, T, D> {
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+impl<K, V, T, D> ReadHandle<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Codec64,
+    D: Semigroup + Codec64,
+{
+    // This handle's since frontier, not the global collection-level one.
+    pub fn since(&self) -> &Antichain<T> {
+        todo!()
+    }
+
+    // NB new_upper of the empty antichain "finishes" this collection, promising
+    // that no more data will ever be read by this handle.
+    //
+    // WIP AntichainRef?
+    pub fn downgrade_since(&self, new_since: Antichain<T>) {
+        todo!("{:?}", new_since)
+    }
+
+    // WIP: The impl is far far easier if this is something we keep single
+    // process at first (as opposed to being able to shard it like snapshots).
+    // Do we think that's fine for short term (i.e. initial platform launch)? We
+    // can definitely revisit this later if it ends up being a bottleneck.
+    //
+    // WIP: This is an opportunity to push down Projection information to save
+    // bandwidth. My thinking so far is that Map and Filter should stay
+    // STORAGE-level concerns.
+    //
+    // WIP: My first attempt here was to return `impl Stream` but that didn't
+    // compile with `todo!`. I don't yet have an opinion on which is better.
+    pub async fn tail(as_of: Antichain<T>) -> Result<Tail<K, V, T, D>, Error> {
+        todo!("{:?}", as_of);
+    }
+
+    // NB the len of the returned vec is num_shards
+    //
+    // WIP This is an opportunity to push down Projection information to save
+    // bandwidth. My thinking so far is that Map and Filter should stay
+    // STORAGE-level concerns.
+    pub async fn snapshot(
+        &self,
+        as_of: Antichain<T>,
+        num_shards: NonZeroUsize,
```

Comment:
What is a snapshot shard? A STORAGE collection would be composed of renditions which have a specific number of shard. Each rendition shard is backed by a single persist collection.

When a STORAGE collection is replayed it will be replayed to as many workers as there are shards. If it makes your life easier we don't need `persist` itself to figure out how to shard a persist collection to an arbitrary number of readers

Reply:
see my response to your other question on snapshot shards